File: Cpp/208_Implement_Trie_(Prefix_Tree).cpp
class Trie {
public:
    /** Initialize your data structure here. */
    Trie() : root(new TrieNode()){
        
    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        TrieNode* node = root;
        for(auto& ch : word){
            int index = ch -'a';
            if(node->children[index] == nullptr){
                node->children[index] = new TrieNode();
            }
            node = node->children[index];
        }
        node->is_leaf = true;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        TrieNode* node = root;
        for(auto& ch : word){
            int index = ch-'a';
            if(node->children[index] == nullptr) return false;
            node = node->children[index];
        }
        return node->is_leaf;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        TrieNode* node = root;
        for(auto& ch : prefix){
            int index = ch-'a';
            if(node->children[index] == nullptr) return false;
            node = node->children[index];
        }
        return true;
    }

private:
    struct TrieNode {
        bool is_leaf;
        vector<TrieNode*> children;

        TrieNode() : is_leaf(false), children(26, nullptr){
        }
    };
    
    TrieNode* root;
    
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: Kotlin/208_Implement_Trie_(Prefix_Tree).kt
class TrieNode {
    val children = Array<TrieNode?>(26) { null }
    var isLeaf = false
}

class Trie() {
    val root = TrieNode()

    fun insert(word: String) {
        var node = root

        for (w in word) {
            val i = w - 'a'
            if (node.children[i] == null) node.children[i] = TrieNode()
            node = node.children[i]!!
        }
        node.isLeaf = true
    }

    fun search(word: String): Boolean {
        var node = root
    
        for (ch in word) {
            val i = ch - 'a'
            if (node.children[i] == null) return false
            node = node.children[i]!!
        }
        return node.isLeaf
    }

    fun startsWith(prefix: String): Boolean {
        var node = root
    
        for (ch in prefix) {
            val i = ch - 'a'
            if(node.children[i] == null) return false
            node = node.children[i]!!
        }

        return true
    }

}

/**
 * Your Trie object will be instantiated and called as such:
 * var obj = Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */